python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java

```diff
@@ -19,93 +19,142 @@
 import org.sonar.check.Rule;
 import org.sonar.python.PythonCheck;
-import org.sonar.python.api.PythonGrammar;
-import org.sonar.python.api.PythonPunctuator;
 import org.sonar.squidbridge.annotations.ActivatedByDefault;
 import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
-import org.sonar.sslr.ast.AstSelect;
+
+import com.sonar.sslr.api.AstAndTokenVisitor;
+import com.sonar.sslr.api.AstNode;
+import com.sonar.sslr.api.Token;
 
 @Rule(
-    key = UselessParenthesisCheck.CHECK_KEY,
-    priority = Priority.MAJOR,
-    name = "Useless parentheses around expressions should be removed to prevent any misunderstanding",
-    tags = Tags.CONFUSING
-)
+  key = UselessParenthesisCheck.CHECK_KEY,
+  priority = Priority.MAJOR,
+  name = "Redundant parentheses around expressions should be removed",
+  tags = Tags.CONFUSING)
 @SqaleConstantRemediation("1min")
 @ActivatedByDefault
-public class UselessParenthesisCheck extends PythonCheck {
+public class UselessParenthesisCheck extends PythonCheck implements AstAndTokenVisitor {
+
   public static final String CHECK_KEY = "S1110";
 
+  private static final String MESSAGE = "Remove those useless parentheses.";
+
+  private Deque<LeftParenthensisData> leftParenthesisDataStack = new LinkedList<>();
+//  private LinkedList<LeftParenthensisData> leftParenthesisDataStack = new LinkedList<>();
+  
+  private LeftParenthensisData last;
+  
+  private Deque<Integer> pairsByLevel = new LinkedList<>();
+
+  private boolean previousTokenWasLeftParenthesis;
+
+  private boolean previousTokenWasRightParenthesis;
+
   @Override
-  public void init() {
-    subscribeTo(
-        PythonGrammar.TEST,
-        PythonGrammar.EXPR,
-        PythonGrammar.NOT_TEST
-    );
+  public void visitFile(AstNode astNode) {
+    leftParenthesisDataStack.clear();
+//    pairsByLevel.clear();
+    previousTokenWasLeftParenthesis = false;
+    previousTokenWasRightParenthesis = false;
   }
 
   @Override
-  public void visitNode(AstNode node) {
-    if (node.is(PythonGrammar.NOT_TEST)) {
-      visitNotTest(node);
-    } else if (node.getNumberOfChildren() == 1 && node.getFirstChild().is(PythonGrammar.ATOM)) {
-
-      AstNode atom = node.getFirstChild();
-      AstSelect selectedParent = atom.select().parent().parent().parent();
-      if (selectedParent.size() == 1) {
-        AstNode parent = selectedParent.get(0);
-        if (isKeywordException(parent)) {
-          checkAtom(atom, true);
-          return;
-        }
+  public void visitToken(Token token) {
+    if (token.getType() == LPARENTHESIS) {
+      if (previousTokenWasLeftParenthesis) {
+        leftParenthesisDataStack.peek().setContiguousLeftParenthesis(token);
       }
+      leftParenthesisDataStack.push(new LeftParenthensisData(token));
+      previousTokenWasLeftParenthesis = true;
+      previousTokenWasRightParenthesis = false;
+
+    } else if (token.getType() == RPARENTHESIS) {
+      checkRedundantParentheses(token);
+//      cleanUpPrevious();
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') This block of commented-out lines of code should be removed. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine)